### DISCUSSION TITLE
What would be the best option for Next.js to save session information for a chatbot?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#57418] [Help] [answered]
### Summary

I need to save some browser information for a chatbot and keep some memory to be fed to the bot...

What about using LocalStorage from the browser? 

Some points to consider is that Next.js might not work all that well with the browser..

Since I don't want to go on try and error for everything.... I  am asking for advise on what would be the best tools to work for Next.js  to keep some chat history fed to the bot and know that you are communicating with an specific browser....  ???

Any feedback on this would be awesome...  thank you!

**Comment by icyJoseph** [comment, marked as answer]
> Some points to consider is that Next.js might not work all that well with the browser..

That's not necessarily true. What happens is that far too often we forget that React rendering should be pure, aka, no side-effects, and reading from the local storage as you render, is not a good idea if we want to keep things pure.

You need to synchronize with the local storage, through useEffects (for completeness sake people might talk about useSyncExternalStore, but keep it simple for now)

So it is perfectly fine to store stuff there.

Do consider that:

- you may accidentally store badly shaped data
- users can manipulate the data, and possibly trick your app
- local storage might not be available, because of a user's privacy settings

You can also use session storage, but that is more volatile. 

- you initialise state to an empty state, say null
- then inside a use-effect with an empty array of dependencies,
  - read from the local storage
  - parse it from JSON into an object or some other useful data type for your case
  - do some sanity checks on the data (validation)
  - set state to that value if its valid, or keep it null otherwise
- as the user interacts with your application, use handlers, callbacks, or use-effects (to a lesser extend) to set the state and save to local storage in one go
